examples/actix-web.rs

```diff
@@ -0,0 +1,103 @@
+use std::sync::Mutex;
+
+use actix_web::{web, App, HttpResponse, HttpServer, Responder, Result};
+use prometheus_client::encoding::text::{encode, Encode};
+use prometheus_client::metrics::counter::Counter;
+use prometheus_client::metrics::family::Family;
+use prometheus_client::registry::Registry;
+
+#[derive(Clone, Hash, PartialEq, Eq, Encode)]
+pub enum RequestType {
+    Get,
+    Post,
+}
+
+#[derive(Clone, Hash, PartialEq, Eq, Encode)]
+pub struct RequestTypeLabels {
+    pub request_type: RequestType,
+}
+
+/// Holds all metrics.
+/// We shouldn't store metrics inside a MetricsCollector -
+/// overwise we would have to take a lock for each increment of any metric
+pub struct Metrics {
+    requests: Family<RequestTypeLabels, Counter>,
+}
+
+impl Metrics {
+    pub fn new() -> Self {
+        Self {
+            requests: Family::default(),
+        }
+    }
+
+    pub fn inc_requests(&self, request_type: RequestType) {
+        self.requests
+            .get_or_create(&RequestTypeLabels {
+                request_type: request_type,
+            })
+            .inc();
+    }
+}
+
+/// Registers and collects metrics
+pub struct MetricsCollector {
+    registry: Registry,
+}
+
+impl MetricsCollector {
+    pub fn new(metrics: &Metrics) -> Self {
+        let mut collector = Self {
+            registry: Registry::default(),
+        };
+        collector.register_metrics(metrics);
+        collector
+    }
+
+    fn register_metrics(&mut self, metrics: &Metrics) {
+        self.registry.register(
+            "requests",
+            "Count of requests",
+            Box::new(metrics.requests.clone()),
+        );
+    }
+
+    pub fn collect(&self) -> Result<String, std::io::Error> {
+        let mut buf = Vec::new();
+        encode(&mut buf, &self.registry)?;
+
+        let buf_str = std::str::from_utf8(buf.as_slice()).unwrap().to_string();
+        Ok(buf_str)
+    }
+}
+
+pub async fn metrics_handler(
+    metrics_collector: web::Data<Mutex<MetricsCollector>>,
+) -> Result<HttpResponse> {
+    // TODO: find the way without locking the mutex
+    let body: String = metrics_collector.lock().unwrap().collect()?;
+    Ok(HttpResponse::Ok().body(body))
+}
+
+pub async fn some_handler(metrics: web::Data<Metrics>) -> impl Responder {
+    metrics.inc_requests(RequestType::Get);
+    format!("okay")
+}
+
+#[actix_web::main]
+async fn main() -> std::io::Result<()> {
+    let metrics = web::Data::new(Metrics::new());
+    // We have to wrap it with Mutex because `Registry` doesn't implement Copy-trait
+    let metrics_collector = web::Data::new(Mutex::new(MetricsCollector::new(&metrics)));
```

Comment:
Note that `encode` takes an immutable reference to `Registry`. Thus there is no need for a `Mutex` in case you never register new metrics after the first Prometheus scrape. An `Arc` should suffice.

Reply:
Reasonable, please check out a [comment](https://github.com/prometheus/client_rust/issues/53#issuecomment-1079764522)